src/input.rs

```diff
@@ -70,6 +70,48 @@ where
             changed_at: ChangedAt::Revision(Revision::ZERO),
         })
     }
+
+    fn set_common(&self, key: &Q::Key, value: Q::Value, changed_at: impl FnOnce() -> ChangedAt) {
+        let map = self.map.upgradable_read();
+
+        // If this value was previously stored, check if this is an
+        // *actual change* before we do anything.
+        if let Some(old_value) = map.get(key) {
+            if old_value.value == value {
+                return;
+            }
+        }
+
+        let key = key.clone();
+
+        let mut map = RwLockUpgradableReadGuard::upgrade(map);
+
+        // Do this *after* we acquire the lock, so that we are not
+        // racing with somebody else to modify this same cell.
+        // (Otherwise, someone else might write a *newer* revision
+        // into the same cell while we block on the lock.)
+        let changed_at = changed_at();
+        let stamped_value = StampedValue { value, changed_at };
+
+        match map.entry(key) {
+            Entry::Occupied(mut entry) => {
+                assert!(
+                    !entry.get().changed_at.is_constant(),
+                    "modifying `{:?}({:?})`, which was previously marked as constant (old value `{:?}`, new value `{:?}`)",
```

Comment:
So, if want to change a constant (say, the user has run a `cargo update`), we need to tear-down runtime completely and re-insert all the inputs, right? An alternative would be to just clean all the derived storages.

Reply:
I wonder if we’ll need some kind of batch modification API for that, to set several constants in one go, bit clean the store only once.

Reply:
Hmm, interesting. It feels like it'd be nice in general to make it so that

```rust
input.set(k1, v1);
input.set(k2, v2);
```

was smart enough not to increment the revision twice. The idea would be to track something like whether the revision is "unobserved" (no derived queries are based on it).

Reply:
Seems like we should file an issue on it.

Reply:
Thinking more about this, seems silly. We should just handle overflow some day if it ever happens. Until then, increments are free =)